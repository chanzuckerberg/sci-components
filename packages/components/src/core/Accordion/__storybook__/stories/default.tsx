import { Args } from "@storybook/react";
import RawAccordion, {
  AccordionDetails,
  AccordionHeader,
} from "src/core/Accordion";
import { EXTRA_SHORT_LOREM_IPSUM } from "src/common/storybook/loremIpsum";

export const Accordion = (props: Args): JSX.Element => {
  const { id, subtitle, useDivider, togglePosition } = props;
  return (
    <RawAccordion
      id={id}
      useDivider={useDivider}
      togglePosition={togglePosition}
      {...props}
    >
      <AccordionHeader id={`${id}-header`} subtitle={subtitle}>
        Accordion Header
      </AccordionHeader>
      <AccordionDetails>{EXTRA_SHORT_LOREM_IPSUM}</AccordionDetails>
    </RawAccordion>
  );
};
