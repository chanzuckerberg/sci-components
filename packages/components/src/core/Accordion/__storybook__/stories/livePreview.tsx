import { Args } from "@storybook/react";
import { Accordion } from "./default";

const livePreviewStyles = {
  display: "grid",
  gridColumnGap: "24px",
  gridRowGap: "50px",
  gridTemplateColumns: "repeat(4, 200px)",
};

export function LivePreviewDemo(props: Args): JSX.Element {
  return (
    <div style={livePreviewStyles as React.CSSProperties}>
      <Accordion
        id="accordion-1"
        togglePosition="right"
        useDivider
        {...props}
      />
      <Accordion
        id="accordion-2"
        togglePosition="right"
        subtitle="Optional Subtitle"
        useDivider
        {...props}
      />
      <Accordion id="accordion-3" togglePosition="left" useDivider {...props} />
      <Accordion
        id="accordion-4"
        togglePosition="left"
        subtitle="Optional Subtitle"
        useDivider
        {...props}
      />
    </div>
  );
}
