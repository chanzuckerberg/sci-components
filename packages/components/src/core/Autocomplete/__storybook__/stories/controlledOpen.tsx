import { Args } from "@storybook/types";
import { DefaultAutocompleteOption } from "src/core/Autocomplete/components/AutocompleteBase";
import { AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS } from "src/common/storybook/AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS";
import { useState } from "react";
import RawAutocomplete from "src/core/Autocomplete";
import { AUTOCOMPLETE_WRAPPER_STYLES } from "../constants";

export const ControlledOpenDemo = <T extends DefaultAutocompleteOption>(
  props: Args
): JSX.Element => {
  const {
    label,
    multiple,
    options = AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS,
  } = props;

  const [open, setOpen] = useState(true);

  return (
    <div style={AUTOCOMPLETE_WRAPPER_STYLES}>
      <RawAutocomplete
        onClickAway={() => {
          setOpen(false);
        }}
        onClick={() => {
          if (open) {
            setOpen(false);
          } else {
            setOpen(true);
          }
        }}
        open={open}
        id="autocomplete-demo"
        disableCloseOnSelect={multiple}
        label={label}
        multiple={multiple}
        options={options}
        getOptionDisabled={(option: T) => {
          return option.name === "Type: feature request";
        }}
        {...props}
      />
    </div>
  );
};
