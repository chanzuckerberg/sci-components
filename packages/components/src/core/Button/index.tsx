import { ButtonProps as RawButtonProps } from "@mui/material";
import React, { ForwardedRef } from "react";
import {
  SDSWarningTypes,
  showWarningIfFirstOccurence,
} from "src/common/warnings";
import { StyledButton, StyledButtonLegacy, StyledMinimalButton } from "./style";
import ButtonIcon from "../ButtonIcon";
import { IconNameToSizes, IconProps } from "../Icon";
import { filterProps } from "src/common/utils";

type ButtonType = "primary" | "secondary" | "tertiary" | "destructive";
type ButtonSize = "small" | "medium" | "large";

type SdsProps =
  | {
      sdsStyle?: "icon";
      sdsType?: "primary" | "secondary" | "tertiary";
      isAllCaps?: boolean;
      isRounded?: boolean;
      sdsSize?: ButtonSize;
      icon?: keyof IconNameToSizes | React.ReactElement<CustomSVGProps>;
      sdsIconProps?: Partial<IconProps<keyof IconNameToSizes>>;
    }
  | {
      sdsStyle?: "minimal";
      sdsType?: "primary" | "secondary";
      isAllCaps?: boolean;
      isRounded?: boolean;
    }
  | {
      sdsStyle?: "rounded" | "square";
      sdsType?: "primary" | "secondary" | "destructive";
      isAllCaps?: boolean;
      isRounded?: boolean;
    };

export type ButtonProps = RawButtonProps & SdsProps;

/**
 * (masoudmanson): The MUI variant prop is determined by the sdsType prop.
 */
const MUI_VARIANT_MAP: { [key in ButtonType]: ButtonProps["variant"] } = {
  destructive: "contained",
  primary: "contained",
  secondary: "outlined",
  tertiary: "text",
};

/**
 * @see https://mui.com/material-ui/react-button/
 */
const Button = React.forwardRef(
  (
    props: ButtonProps,
    ref: ForwardedRef<HTMLButtonElement>
  ): JSX.Element | null => {
    const { sdsStyle, sdsType } = props;

    if (!sdsStyle || !sdsType) {
      showWarningIfFirstOccurence(SDSWarningTypes.ButtonMissingSDSProps);
    }

    if (typeof props?.isAllCaps === "boolean" && sdsStyle !== "minimal") {
      showWarningIfFirstOccurence(SDSWarningTypes.ButtonMinimalIsAllCaps);
    }

    // isAllCaps is only used for the Minimal Button.
    // It defaults to true for the Minimal Button.
    // It defaults to false for all other buttons.
    const isAllCaps =
      typeof props?.isAllCaps === "boolean"
        ? props?.isAllCaps
        : sdsStyle === "minimal";

    type PropsWithDefaultsType = ButtonProps & { isAllCaps?: boolean };
    const propsWithDefault: PropsWithDefaultsType = { ...props, isAllCaps };

    switch (true) {
      case sdsStyle === "icon":
        if (props?.icon !== undefined) {
          // (masoudmanson): We need to remove the props that are not supported by
          // the ButtonIcon component.
          const excludedProps: (keyof PropsWithDefaultsType)[] = [
            "startIcon",
            "endIcon",
            "sdsStyle",
            "isAllCaps",
            "sdsType",
          ];
          const finalProps = filterProps(propsWithDefault, excludedProps);

          return (
            <ButtonIcon
              icon={props?.icon}
              {...finalProps}
              sdsType={sdsType as Exclude<ButtonType, "destructive">}
              ref={ref}
            />
          );
        } else {
          showWarningIfFirstOccurence(
            SDSWarningTypes.ButtonIconMissingIconProp
          );
          return null;
        }

      case sdsStyle === "minimal":
        return (
          <StyledMinimalButton
            color="primary"
            variant="text"
            {...propsWithDefault}
            ref={ref}
          />
        );

      case sdsStyle === "rounded":
      case sdsStyle === "square":
        return (
          <StyledButton
            color="primary"
            variant={sdsType ? MUI_VARIANT_MAP[sdsType] : "text"}
            {...propsWithDefault}
            ref={ref}
          />
        );

      default:
        return <StyledButtonLegacy {...propsWithDefault} ref={ref} />;
    }
  }
);

export default Button;
