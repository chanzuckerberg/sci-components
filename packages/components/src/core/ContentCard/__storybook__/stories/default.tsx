import { Args } from "@storybook/types";
import { MEDIUM_LOREM_IPSUM } from "src/common/storybook/loremIpsum";
import Button from "src/core/Button";
import RawContentCard, {
  ContentCardActions,
  ContentCardBody,
} from "src/core/ContentCard";

export const ContentCard = (props: Args): JSX.Element => {
  const { buttons, visualElementType, ...rest } = props;

  return (
    <RawContentCard visualElementType={visualElementType} {...rest}>
      <ContentCardBody>{MEDIUM_LOREM_IPSUM}</ContentCardBody>

      <ContentCardActions>
        {buttons && buttons.map((button: typeof Button) => button)}
      </ContentCardActions>
    </RawContentCard>
  );
};
