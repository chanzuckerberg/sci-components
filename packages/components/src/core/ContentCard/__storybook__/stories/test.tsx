import { Args } from "@storybook/react-webpack5";
import { MEDIUM_LOREM_IPSUM } from "src/common/storybook/loremIpsum";
import Button from "src/core/Button";
import ContentCard, {
  ContentCardActions,
  ContentCardBody,
} from "src/core/ContentCard";
import Icon from "src/core/Icon";

export const TestDemo = (props: Args): JSX.Element => {
  return (
    <ContentCard
      data-testid="content-card"
      visualElementType="icon"
      icon={<Icon sdsIcon="Compass" sdsSize="xl" />}
      titleText="Content Card Title"
      subtitleText="Content Card Subtitle"
      metadataText="Content Card Metadata"
      {...props}
    >
      <ContentCardBody>{MEDIUM_LOREM_IPSUM}</ContentCardBody>
      <ContentCardActions>
        <Button sdsStyle="minimal" sdsType="primary">
          Button
        </Button>
      </ContentCardActions>
    </ContentCard>
  );
};
