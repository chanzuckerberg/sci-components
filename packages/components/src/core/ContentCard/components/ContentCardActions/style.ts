import styled from "@emotion/styled";
import { CardActions, CardActionsProps } from "@mui/material";
import { ButtonProps } from "src/core/Button";
import { CommonThemeProps, getSpaces } from "src/core/styles";

const doNotForwardProps = ["buttonsPosition"];

export type StyledCardActionsProps = CardActionsProps &
  CommonThemeProps & {
    buttonsPosition?: "left" | "right";
    children?: Array<React.ReactElement<ButtonProps>>;
  };

export const StyledCardActions = styled(CardActions, {
  shouldForwardProp: (prop: string) => !doNotForwardProps.includes(prop),
})`
  ${(props: StyledCardActionsProps) => {
    const { buttonsPosition = "left" } = props;

    const spaces = getSpaces(props);

    return `
      margin: 0;
      padding: ${spaces?.xl}px 0 0;
      display: flex;
      gap: ${spaces?.m}px;
      justify-content: ${buttonsPosition === "right" ? "flex-end" : "flex-start"};

      & > * {
        margin: 0 !important;
      }
    `;
  }}
`;
