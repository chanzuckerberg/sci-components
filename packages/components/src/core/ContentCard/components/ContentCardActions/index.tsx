import React, { forwardRef, ReactElement, ReactNode } from "react";
import { StyledCardActions } from "./style";
import Button, { ButtonProps } from "src/core/Button";
import {
  SDSWarningTypes,
  showWarningIfFirstOccurence,
} from "src/common/warnings";
import { ContentCardProps } from "../..";

export interface ContentCardActionsProps {
  buttonsPosition?: "left" | "right";
  clickableCard?: ContentCardProps["clickableCard"];
  classes?: {
    cardActions?: string;
  };
  children:
    | React.ReactElement<ButtonProps>
    | Array<React.ReactElement<ButtonProps>>;
}

const isButtonElement = (
  child: ReactNode
): child is ReactElement<ButtonProps> => {
  if (React.isValidElement(child) && child.type === Button) {
    return true;
  } else {
    showWarningIfFirstOccurence(SDSWarningTypes.ContentCardActionsOnlyButtons);
    return false;
  }
};

/**
 * @see https://mui.com/material-ui/api/card-actions/
 */
const ContentCardActions = forwardRef<HTMLDivElement, ContentCardActionsProps>(
  function ContentCardActions(
    props: ContentCardActionsProps,
    ref
  ): JSX.Element | null {
    const { buttonsPosition, clickableCard, children, classes } = props;

    /**
     * (masoudmanson):
     * We need to ensure that only SDS buttons are used within the
     * ContentCardActions component slot. This is to prevent any potential
     * issues with using other components within the ContentCardActions component slot.
     */
    const validChildren = React.Children.toArray(
      Array.isArray(children) ? children : [children]
    ).filter(isButtonElement);

    /**
     * If the card is clickable, it acts as a `<button>` itself and cannot
     * have `<button>` elements as children. If multiple buttons are provided,
     * only the first one will be kept, and a warning will be shown.
     *
     * Any additional buttons cannot remain as `<button>` elements due to
     * DOM hierarchy constraints; they will be converted to `<a>` tags instead.
     */
    if (clickableCard) {
      if (validChildren.length > 1) {
        showWarningIfFirstOccurence(
          SDSWarningTypes.ClickableContentCardNumberOfButtons
        );
      }

      /**
       * (masoudmanson): Since a clickable card can have at most one button,
       * we select the first button from the validChildren array and ignore the rest.
       */
      const clickableCardButton = validChildren.shift();

      if (!clickableCardButton) {
        return null;
      }

      return (
        <StyledCardActions
          ref={ref}
          buttonsPosition={buttonsPosition}
          className={classes?.cardActions}
        >
          {[
            React.cloneElement(
              clickableCardButton as React.ReactElement<ButtonProps>,
              {
                component: "div",
                ...(clickableCardButton?.props as ButtonProps),
              }
            ),
          ]}
        </StyledCardActions>
      );
    }

    return (
      <StyledCardActions
        ref={ref}
        buttonsPosition={buttonsPosition}
        className={classes?.cardActions}
      >
        {validChildren}
      </StyledCardActions>
    );
  }
);

export default ContentCardActions;
