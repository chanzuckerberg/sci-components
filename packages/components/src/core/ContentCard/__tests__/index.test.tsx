import { generateSnapshots } from "@chanzuckerberg/story-utils";
import { composeStories } from "@storybook/react-webpack5";
import { fireEvent, render, screen } from "@testing-library/react";
import * as stories from "../__storybook__/index.stories";
import Button from "src/core/Button";
import ContentCardActions from "../components/ContentCardActions";
import { ContentCardBody } from "..";

global.ResizeObserver = class {
  observe() {}

  unobserve() {}

  disconnect() {}
};

// Returns a component that already contain all decorators from story level, meta level and global level.
const { Test } = composeStories(stories);
const CONTENT_CARD_TEST_ID = "content-card";

describe("ContentCard", () => {
  generateSnapshots(stories);

  it("renders content card with default props", () => {
    render(<Test />);
    expect(screen.getByTestId(CONTENT_CARD_TEST_ID)).toBeInTheDocument();
  });

  it("renders content card with image", () => {
    render(
      <Test
        visualElementType="image"
        image="https://picsum.photos/1000"
        imageSize={300}
      />
    );
    const image = screen.getByRole("img");
    expect(image).toBeInTheDocument();
  });

  it("renders content card with icon", () => {
    render(
      <Test
        visualElementType="icon"
        icon={<div data-testid="test-icon">Icon</div>}
      />
    );
    expect(screen.getByTestId("test-icon")).toBeInTheDocument();
  });

  it("renders content card with text content", () => {
    render(
      <Test
        overlineText="Overline"
        titleText="Title"
        subtitleText="Subtitle"
        metadataText="Metadata"
      />
    );
    expect(screen.getByText("Overline")).toBeInTheDocument();
    expect(screen.getByText("Title")).toBeInTheDocument();
    expect(screen.getByText("Subtitle")).toBeInTheDocument();
    expect(screen.getByText("Metadata")).toBeInTheDocument();
  });

  it("renders clickable card", () => {
    const onClick = jest.fn();
    render(<Test clickableCard onClick={onClick} />);
    const card = screen.getByTestId(CONTENT_CARD_TEST_ID);
    fireEvent.click(card);
    expect(onClick).toHaveBeenCalled();
  });

  it("renders content card with buttons", () => {
    render(
      <Test>
        <ContentCardBody>
          <div>Content Card Body</div>
        </ContentCardBody>

        <ContentCardActions>
          <Button key="1" sdsStyle="minimal" sdsType="primary">
            Button
          </Button>
          <Button key="2" sdsStyle="minimal" sdsType="secondary">
            Second Button
          </Button>
        </ContentCardActions>
      </Test>
    );
    expect(screen.getByText("Button")).toBeInTheDocument();
  });

  it("renders content card with different button positions", () => {
    render(
      <Test buttonsPosition="right">
        <ContentCardBody>
          <div>Content Card Body</div>
        </ContentCardBody>

        <ContentCardActions>
          <Button key="1" sdsStyle="minimal" sdsType="primary">
            Button
          </Button>
          <Button key="2" sdsStyle="minimal" sdsType="secondary">
            Second Button
          </Button>
        </ContentCardActions>
      </Test>
    );
    const button = screen.getByText("Button");
    expect(button).toBeInTheDocument();
  });

  it("renders content card with different image positions", () => {
    render(
      <Test
        visualElementType="image"
        image="https://picsum.photos/1000"
        imagePosition="right"
      />
    );
    const image = screen.getByRole("img");
    expect(image).toBeInTheDocument();
  });

  it("renders content card with different types", () => {
    render(<Test sdsType="narrow" />);
    expect(screen.getByTestId(CONTENT_CARD_TEST_ID)).toBeInTheDocument();
  });

  it("renders content card with ReactNode as titleText", () => {
    const CustomTitle = () => (
      <span data-testid="custom-title">Custom Title</span>
    );
    render(<Test titleText={<CustomTitle />} />);
    expect(screen.getByTestId("custom-title")).toBeInTheDocument();
    expect(screen.getByText("Custom Title")).toBeInTheDocument();
  });
});
