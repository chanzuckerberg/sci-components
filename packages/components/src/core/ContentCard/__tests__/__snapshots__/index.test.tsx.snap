// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentCard Default story renders snapshot 1`] = `
<div
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-ok5j5u-MuiPaper-root-MuiCard-root"
>
  <div
    class="css-20z69m"
  >
    <img
      alt="Content Card Media"
      class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-fq4hyx-MuiCardMedia-root"
      src="https://picsum.photos/1000"
    />
  </div>
  <div
    class="css-1wmo34t"
  >
    <div>
      <div
        class="css-1jlejkr"
      >
        <p
          class="css-5sjq7l"
        >
          Overline Text
        </p>
        <p
          class="css-5sjq7l"
        >
          Title Text
        </p>
        <p
          class="css-5sjq7l"
        >
          Subtitle Text
        </p>
      </div>
      <p
        class="css-16dysem"
      >
        Metadata Text
      </p>
    </div>
    <div
      class="css-6kyaea"
    >
      <div
        class="MuiCardContent-root css-1ybx2ee-MuiCardContent-root"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Imperdiet massa tincidunt nunc pulvinar sapien et ligula ullamcorper. Urna duis convallis convallis tellus id interdum velit laoreet id. Donec ultrices tincidunt arcu non sodales. Aliquam eleifend mi in nulla posuere.
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing css-1w4zes7-MuiCardActions-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1cs60p5-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Primary Button
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-jrswey-MuiButtonBase-root-MuiButton-root"
          data-style="minimal"
          tabindex="0"
          type="button"
        >
          Minimal Button
          <span
            class="MuiButton-icon MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
          >
            <div
              class="css-1ft6wgl"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-bgw7zo-MuiSvgIcon-root"
                data-file-name="IconChevronRightSmall"
                data-testid="IconChevronRightSmall"
                fillcontrast="white"
                focusable="false"
                height="16"
                viewBox="0 0 16 16"
                width="16"
              />
            </div>
          </span>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
